### PR TITLE
Arm64 Support In Dockerfile

### PR DESCRIPTION
**Pull request by odidev**
#1

Added the support for arm64 in Dockerfile.
As GHC official binaries are not yet released for arm64, I am building the GHC binaries from its source.

### CODE REVIEW

**Comment by odidev** [comment]
Ping @psftw could you please share your thoughts on this.

**Comment by odidev** [comment]
@psftw Kindly have a look on this PR

**Comment by psftw** [comment]
Thanks for your work on this @odidev. Unfortunately right now I don't have the time to see this through, and merging/publishing prematurely would be a mistake. This PR only addresses 8.8, which in the past I had experimented with and ran in to usability problems (many of which may be resolved now). I'm not sure of the current 8.8/8.10 status for GHC (+ tooling/ecosystem) on ARM, but I'd like to do significant testing and research to make sure we are not releasing anything we shouldn't be. I'm pretty sure a number of caveats would need to be documented for users in any case. I'd also like to have a sustainable smoke-testing process that can integrate with (or replace) the existing github actions. It is possible to temporarily spin up an ARM server and use it as a github action runner, but this is not ideal.

**Comment by GrahamCampbell** [comment]
I think this needs doing for 8.10 too?

**Comment by odidev** [comment]
Hi,

Added the support for arm64 in Dockerfile for 8.8 and 8.10
8.8.4 source for arm64 is not available so i am using 8.8.3 for arm64.
GHC official support for arm64 is only available for 8.8 stretch. I have cross compiled the 8.8 stretch source in buster and it is working.
GHC official support for arm64 is not available for 8.10 stretch. Hence I have only added the support for ghc 8.10 buster.

The stack consist of binary whereas binary of ghc is unavailable so i cannot verify the tar as these tar consists of the source for arm64.

**Comment by odidev** [comment]
@GrahamCampbell 
can you please review and let me know if any changes required?

**Comment by GrahamCampbell** [comment]
I think the best approach now would be to only add ARM support for GHC 9.0, and use the official builds, which will be available, and working properly. 8.x series should be left alone.

**Comment by GrahamCampbell** [comment]
https://downloads.haskell.org/~ghc/8.10.3/ghc-8.10.3-aarch64-deb10-linux.tar.xz

File: 8.10/buster/Dockerfile [hunk 1]
```diff
@@ -6,41 +6,56 @@ RUN apt-get update && \
...
-    gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys ${STACK_KEY} && \
-    gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys ${STACK_RELEASE_KEY} && \
-    curl -fSL https://github.com/commercialhaskell/stack/releases/download/v${STACK}/stack-${STACK}-linux-x86_64.tar.gz -o stack.tar.gz && \
-    curl -fSL https://github.com/commercialhaskell/stack/releases/download/v${STACK}/stack-${STACK}-linux-x86_64.tar.gz.asc -o stack.tar.gz.asc && \
-    gpg --batch --trusted-key 0x575159689BEFB442 --verify stack.tar.gz.asc stack.tar.gz && \
-    tar -xf stack.tar.gz -C /usr/local/bin --strip-components=1 && \
+RUN arch=`uname -m` && \
+    export GNUPGHOME="$(mktemp -d)" && \
+    if [ $arch  = "aarch64" ] ; then \
+        curl -fSL https://github.com/commercialhaskell/stack/releases/download/v2.1.3/stack-2.1.3-linux-$arch.tar.gz -o  stack.tar.gz && \
```
Comment (GrahamCampbell) [inline 8.10/buster/Dockerfile:45]: version should not be hard coded

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by odidev** [comment]
I have amended the changes as per the given suggestions.

Now this PR only consists of the changes for 8.10 buster, here **8.10.3** ghc version is used for arm64 development and the latest version of stack available for arm64 is **2.1.3**. So I have updated the stack version for arm64. 

Can you please review the latest commits and let me know if any further changes is required

**Comment by GrahamCampbell** [comment]
Not up to me at this point. Up to one of the maintainers with write permissions.

**Comment by GrahamCampbell** [comment]
Likely they will say wait till the debian builds are ready, then change the code back: https://github.com/haskell/docker-haskell/issues/31#issuecomment-751467630

**Comment by psftw** [comment]
I saw there was a GHC 8.10.4 released yesterday, so I made a few tweaks to this PR and built it on an AWS Graviton system. While I was able to get it "working", there are still critical issues with releasing an aarch64 image. Without cabal-install (and ideally Stack as well) I don't think we should ship. I would rather the issues with packaging get resolved upstream instead of hacking something here. The line in your PR to install the cabal-install package brings in another set of packages including an older GHC from upstream Debian repositories which is not what we want. Similarly, including an old version of Stack because they no longer provide binaries for aarch64 is also not going to fly.

**Comment by hasufell** [comment]
ghcup supports arm now and I also have an unofficial stack binary: https://downloads.haskell.org/ghcup/unofficial-bindists/stack/2.7.3/stack-2.7.3-linux-armv7 (not distributed... these were built in an attempt to migrate stack CI to GHC gitlab, which has first-class support for ARM)

See:

* https://github.com/haskell/docker-haskell/pull/47
* https://github.com/commercialhaskell/stack/issues/5619
* https://github.com/commercialhaskell/stack/issues/5610

**Comment by Bodigrim** [comment]
@psftw any chance to move this forward? Arm docker images for GHC would be a great help for package maintainers.

**Comment by AlistairB** [comment]
Thanks for the ping on this @Bodigrim . FYI @psftw has stepped down from maintaining this project after many years of service.

I believe the major blockers at this point are:

- stack official ARM releases. It seems that this existed for 2.7.1, but not for 2.7.3. I would be inclined to only include an official stack release.
- CI that can build the images on ARM based CPUs. Actually it seems that travis might support this https://docs.travis-ci.com/user/multi-cpu-architectures/#multi-cpu-availability.

I'll see if I can get something fully working in travis CI with stack 2.7.1 for now. If that works then perhaps we can bug the stack maintainers (or offer a hand) :)

**Comment by AlistairB** [comment]
Hmm dang, seems that the haskell github org is out of travis ci credits.

image

**Comment by Bodigrim** [comment]
Yeah, Travis is no longer an option for OSS maintainers. I have successful experience with ARM machines provided by drone.io, see https://github.com/haskell/bytestring/blob/master/.drone.yml and https://github.com/haskell/text/blob/master/.drone.yml

Even a `cabal`-only docker image would be very helpful.

**Comment by AlistairB** [comment]
Ah drone looks nice thanks I'll try that.

Yeah, ideally we get stack too, but if it is not likely to be possible anytime soon I think it can be omitted.

**Comment by AlistairB** [comment]
Although, I think someone who is a haskell org admin would need to follow https://docs.drone.io/server/provider/github/ or apply the setup bytestring has to docker-haskell?

**Comment by hasufell** [comment]
> Arm docker images for GHC would be a great help for package maintainers.

I've been working on [ghcup docker images](https://gitlab.haskell.org/haskell/ghcup-hs/-/merge_requests/173), which are more flexible and lightweight than `haskell/docker` and would use the GHC CI infrastructure to build. But I haven't gotten around it yet.

**Comment by Bodigrim** [comment]
@AlistairB I set up drone.io for `bytestring` and `text` myself, so you should be able to do the same for `docker-haskee` as soon as you have enough rights for the repo. CC @chessai 
Alternatively you can try emulated GitHub actions environment, e. g., https://github.com/haskell/bytestring/blob/e682cf7ad44c8299aed23e649472d8ab559109d4/.github/workflows/ci.yml#L69-L90

@hasufell true, but installing everything from the scratch takes too long for a low-latency CI, which is my use case.

**Comment by chessai** [comment]
> @AlistairB I set up drone.io for `bytestring` and `text` myself, so you should be able to do the same for `docker-haskell` as soon as you have enough rights for the repo. CC @chessai Alternatively you can try emulated GitHub actions environment, e. g., https://github.com/haskell/bytestring/blob/e682cf7ad44c8299aed23e649472d8ab559109d4/.github/workflows/ci.yml#L69-L90
> 
> @hasufell true, but installing everything from the scratch takes too long for a low-latency CI, which is my use case.

@Bodigrim, @AlistairB already has admin access to this repo (docker-haskell). Did you mean another repo?

**Comment by Bodigrim** [comment]
Ah, good, then he should be able to set up drone.io for `docker-haskell` himself.

**Comment by AlistairB** [comment]
Ah hmm, ok let me have another go. The instructions seemed to suggest it needs to be done at an org level.

**Comment by AlistairB** [comment]
So I've been [playing with the emulated approach](https://github.com/haskell/docker-haskell/pull/53/). The good news I think I have a good pattern now to make the Dockerfiles processor architecture agnostic in that PR. Unfortunately the emulated approach is not working as `uname -m` apparently returns [the processor architecture of the real host on the CI agent](https://github.com/uraimo/run-on-arch-action/issues/21), not that of the emulated ARM64 docker container.

@Bodigrim Re drone, am I correct in thinking I would need to run 'servers' and/or 'runners' myself somewhere? The instructions ask me to set up a GitHub OAuth App.

image

And I would need to host some `http://drone.myserver/login` or something?

I don't really want to host anything myself (if I am understanding this all correctly) and if I did I would probably set up a self hosted github CI agent.

I think if there are no viable non-self hosted options, it is actually not the worst to just have no CI verifying ARM in this repo. As when new releases are PRed against https://github.com/docker-library/official-images as part of the release process their CI should check everything.

**Comment by Bodigrim** [comment]
@AlistairB Drone provides both servers and runners. I remember that GitHub integration was not documented well, but I managed to setup it for bytestring and text - just do not remember how :) 

Dunno, try adding .drone.yml to a fork first and see whether you can integrate it?

**Comment by AlistairB** [comment]
Ok thanks that is good to know. I will figure this out! 😅

**Comment by AlistairB** [comment]
Just an update:

- So it turns out there is a manual approval process for cloud.drone.io. I am waiting for this, then I think I am good to test the updated images on ARM.
- I have pinged the stack people on the HF slack about [2.7.3 missing aarch64 binaries](https://github.com/commercialhaskell/stack/issues/5610) and whether stack intends to produce aarch64 binaries in the future. I'd like to at least get a response from the stack people before releasing something with stack 2.7.1 (which I think is my current preference). Another option is unofficial stack binaries, but I really think the official images should not include anything unofficial.

**Comment by AlistairB** [comment]
- Still waiting for drone.io account.. Another option is try to use haskell gitlab ci, like haskell language server, but that looks really complex to set up so would rather avoid.
- I pinged the stack people and someone put their hand up to automate and add 2.7.3 ARM64 binaries. However, I have investigated and it seem the official images people are happy to have old versions of binaries when newer versions have issues. I am happy to release with 2.7.1 (at least while I see no actual movement to on the 2.7.3 issue) and https://github.com/haskell/docker-haskell/pull/53 has this fallback logic.
- https://github.com/haskell/docker-haskell/pull/53 is ready to go minus CI testing ARM64 on drone.io . Feedback welcome.

**Pull request closed without merging**